qa/L0_model_update/instance_update_test.py

```diff
@@ -0,0 +1,310 @@
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import unittest
+import time
+import concurrent.futures
+import numpy as np
+import tritonclient.grpc as grpcclient
+from tritonclient.utils import InferenceServerException
+from models.instance_init_del.util import *
+
+
+class TestInstanceUpdate(unittest.TestCase):
+
+    def setUp(self):
+        # Reset counters
+        reset_initialize_count()
+        reset_finalize_count()
+        # Reset batching
+        disable_batching()
+        # Reset delays
+        set_initialize_delay(0)
+        set_infer_delay(0)
+        # Initialize client
+        self.__triton = grpcclient.InferenceServerClient("localhost:8001",
+                                                         verbose=True)
+
+    def __get_inputs(self, shape=(2,)):
+        inputs = [grpcclient.InferInput("INPUT0", shape, "FP32")]
+        inputs[0].set_data_from_numpy(np.ones(shape, dtype=np.float32))
+        return inputs
+
+    def test_add_rm_add(self):
+        # Load model
+        update_instance_group("{\ncount: 3\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 3)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Add 1 instance
+        update_instance_group("{\ncount: 4\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 4)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Remove 1 instance
+        update_instance_group("{\ncount: 3\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 4)
+        self.assertEqual(get_finalize_count(), 1)
+        self.__triton.infer("instance_init_del", self.__get_inputs((4,)))
+        # Add 1 instance
+        update_instance_group("{\ncount: 4\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 5)
+        self.assertEqual(get_finalize_count(), 1)
+        self.__triton.infer("instance_init_del", self.__get_inputs((1,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 5)
+        self.assertEqual(get_finalize_count(), 5)
+        with self.assertRaises(InferenceServerException):
+            self.__triton.infer("instance_init_del", self.__get_inputs((5,)))
+
+    def test_rm_add_rm(self):
+        # Load model
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
```

Comment:
Another interesting case is to set the count to 0, send out infer request, there should be no responses as there are no instances available, then increase the instance count which would return some responses.

Reply:
Added a test verifying this behavior: [Add test for zero instance count](https://github.com/triton-inference-server/server/pull/5635/commits/124c594b01b986771b3b7c69df8936be3cc08d43)